### ISSUE TITLE
Can this be tested and run in an internal environment with no access to the internet?

### ISSUE THREAD

**Issue by Pattywaak** [thinkst/canarytokens-docker#70]
Attempting to contruct a project in which a ubuntu virtual machine hosts docker, is there a way to configure it so alerts can be generated and activated without internet connection?

**Comment by jayjb** [comment]
HI @Pattywaak,

So there will be a couple considerations here if you want it to work without access to the internet:
1. triggering incidents requires either a hostname lookup (DNS) or an HTTP request to the canarytokens server. So as long as you have a working DNS server in your internal network that can resolve that hostname of your canarytokens server, you should be able to use the token internally. The problem though would be that to trigger them, you would need to be on that internal network to reach your canarytokens server.

2. alerting. Since the alerting is generally a webhook or some mail service, you would need to ensure that whichever you choose can still receive data from your canarytokens service and can reach out to whichever alert method you choose. So if you decided a webhook, you would need to ensure that your canarytoken server can post a HTTP request to your webhook's address. 

TLDR; it is not impossible, but it does require some extra thinking with regards to triggering and alerting. Please let me know why and what the environment might look like and i can try help.

**Comment by Pattywaak** [comment]
Hey @jayjb 

So the whole environment isn't entirely dependent on not having internet connection. Is there a way that alerts can be sent without providing a domain name. In my current environment I have localhost set up as the domain, I am able to access the front end and generate the tokens at will, however, I am not able to trigger these events no matter what I try. Even tokens such as the microsoft word one don't trigger the tokens and don't send alerts to the provided email address.

Thanks, let me know if there is anything else you need to know in order to answer the question better.

**Comment by jayjb** [comment]
Hi @Pattywaak,

So i think figuring out each issue separately would be useful; i.e. getting the tokens to trigger and then determining which mechanism for alerting you would like.

Would you mind providing me with some of the information in your current config files for your Canarytokens server?

My suspicion is that if you are using localhost as the domain for everything for your Canarytokens server it will mean that your Canarytokens that are generated will something of the form: 
- `<token>.localhost` if it is a DNS Canarytoken
- `localhost/<token>/something.jpg` if it is a HTTP Canarytoken.

The issue with the above is that if you try resolve those URLS, your computer will direct the request to your localhost (and not necessarily to your Canarytokens server). If you are on the same network as your Canarytokens server, you could try tripping a Canarytoken by changing `localhost` to the Canarytoken server IP.

**Comment by Pattywaak** [comment]
Below are the current configurations to the frontend.env and switchboard.env files. I have only made a few changes to the dockercompose.yml file regarding the ports but I don't suspect that being the issue.

Frontend.env

CANARY_DOMAINS=localhost
CANARY_NXDOMAINS=pdftoken.com

#CANARY_GOOGLE_API_KEY=

CANARY_WEB_IMAGE_UPLOAD_PATH=/uploads
#Enter number so 1024*1024*10 = 10485760
CANARY_MAX_UPLOAD_SIZE=10485760

LOG_FILE=frontend.log

Switchboard.env
                                       
#CANARY_MAILGUN_DOMAIN_NAME=
#CANARY_MAILGUN_API_KEY=
#CANARY_MANDRILL_API_KEY=
CANARY_SENDGRID_API_KEY=[sendgrid api key]
CANARY_PUBLIC_IP=[public IP address]
CANARY_PUBLIC_DOMAIN=localhost
CANARY_ALERT_EMAIL_FROM_ADDRESS=[one of my gmail emails]
CANARY_ALERT_EMAIL_FROM_DISPLAY="Example Canarytokens"
CANARY_ALERT_EMAIL_SUBJECT="Canarytokens"
#CANARY_IPINFO_API_KEY=

CANARY_TOKEN_RETURN=fortune

CANARY_WEB_IMAGE_UPLOAD_PATH=/uploads

LOG_FILE=switchboard.log

# Default protocol to use in notification links. If you don't use LetsEncrypt, set to http
PROTOCOL=https

I will experiment with using the IP address that the server uses and get back to you on that.

**Comment by jayjb** [comment]
Hi @Pattywaak,

Thanks for the additional information. Would you mind letting me know about the changes to the docker-compose file, just so i can be certain?

Looking at these environment files you have supplied; using `localhost` as the `CANARY_DOMAINS` address isn't recommended. It means that you will only be able to trigger the Canarytokens you create if you trigger them on that same host. The reason being that that `CANARY_DOMAINS` value needs to be resolved to your Canarytokens Server. 

If any other machine tries to resolve a `localhost` address, it will point to itself and that request will not reach your Canarytokens server (and hence you will not trigger an alert and receive a notification).

Please let me know if this is helpful

**Comment by Pattywaak** [comment]
Hey @jayjb,

I have made a few changes to the server and I have gotten a little further.
Firstly, I have been able to get a webhook request from creating a token but I have not been able to get a webhook post from triggering a token.
Secondly, I have experimented with creating a DNS server on the local network and I am yet to see the affects of that. 
Thirdly, I have had no luck in receiving any emails to the attached address from triggering a token. 

I was wondering how the alerting/triggering process actually works in order to get a better understanding of the process and be able to find a solution. 

Cheers

**Comment by Shango-13** [comment]
Hi, 

I am running a similar setup with a the same goal but i am having issues accessing the front end page  /generate from my mint host.

I have docker installed and running on Ubuntu server in virtual box. I am using a dns server with static ip.

initially when i ran docker-compose up i had errors with the switchboard and nginx ports and I amended them as bellow in the file docker-compose.yml 

switchboard:
    build: ./canarytokens/
    restart: always
    image: thinkst/canarytokens
    ports:
     - "54:53"
     - "54:53/udp"
     - "25:25"

nginx:
    restart: always
    image: thinkst/canarytokens_nginx
    build: ./nginx/
    ports:
     - "81:80"

after this docker all ran in the server but i cannot access http://canary.server/generate from my host browser

my frontend.env is has bellow

CANARY_DOMAINS=canary.server (
#CANARY_NXDOMAINS=

#CANARY_GOOGLE_API_KEY=

CANARY_WEB_IMAGE_UPLOAD_PATH=/uploads
#Enter number so 1024*1024*10 = 10485760
CANARY_MAX_UPLOAD_SIZE=10485760

LOG_FILE=frontend.log

and my switchboard.env

#CANARY_MAILGUN_API_KEY=
#CANARY_MANDRILL_API_KEY=
#CANARY_SENDGRID_API_KEY=
CANARY_PUBLIC_IP=172.17.0.1 (docker ip, had before my dns static ip but with same issue)
CANARY_PUBLIC_DOMAIN=canary.server
#CANARY_ALERT_EMAIL_FROM_ADDRESS=
#CANARY_ALERT_EMAIL_FROM_DISPLAY=
#CANARY_ALERT_EMAIL_SUBJECT=
#CANARY_IPINFO_API_KEY=

CANARY_TOKEN_RETURN=fortune

CANARY_WEB_IMAGE_UPLOAD_PATH=/uploads

LOG_FILE=switchboard.log

# Default protocol to use in notification links. If you don't use LetsEncrypt, set to http
PROTOCOL=https

at this stage i just want to have the front page up before i setup any alerts.

in my server i've tried with ufw disable and still no joy. 
not sure if relevant but i can ssh my server using my static ip.

I have reviewed the github tutorial and ubuntu documentation but i cannot find what i am missing?

can i have some guidance please?


thank you

**Comment by jayjb** [comment]
Hi @Shango-13,

Thanks for giving me so much information. Lets see if we can figure this issue out. 

The first test I would like you to try, is running `docker-compose up` (using your edited `docker-compose.yml`) and then navigating `http://canary.server:81/generate`. The reason i want to try this is because you have changed the port that nginx is listening on (usually it is listening on the default http 80 so you wouldn't have to specify the port). 

The other bit here that I want to make sure about is that the hostname `canary.server` is pointing to your Virtualbox IP. This can be done in a number of ways but somewhere a DNS record needs to point your `canary.server` to your Virtualbox IP that is running canarytokens docker. (an often quick way to get around dealing with DNS for testing, is to edit your `/etc/hosts` file and include, say `canary.server` and point it at the Virtualbox static IP).

Is there anything else that you may be running on that mint VM? Im curious what errors you may have received?

**Comment by Shango-13** [comment]
Hi @jayjb,

Thank you for your help in advance.

Started docker-compose up and tried http://canary.server:81/generate

in Firefox  >  We can’t connect to the server at canary.server.
in Chromium > DNS_PROBE_FINISHED_NXDOMAIN

checking /etc/hosts

127.0.0.1 localhost
127.0.1.1 canary

included the 192.168.0.133 canary.server and still no joy

on my Mint i just have virtual box installed and setup firefox to run sandboxed, hence why tested with chromium also.

ufw disable didn't show any progress.

Because i am a noob, diving head 1st, let me go through what I've done because surely I've messed up on the basics and to save everyone's time.

So i have ubuntu server runing in a vm  with hostname canary. From the tutorials i understood that i needed to created a domain, so i setup canary.server. Also setup fixed ip.

now i am thinking that i should just start from 0 but i have some questions

Goal: Canary working from a VM and I'm only interested and testing the setup between guest and host

In a fresh Ubuntu Server install running on a VM do  I specifically need an Static IP and to setup a specific DNS server?

Or can i just install Server > install Docker > setup Canary tokens?

What basics do you advise for me to review?

many thanks
